software/bsg_manycore_lib/bsg_mcs_mutex.S

```diff
@@ -0,0 +1,80 @@
+// MCS mutex
+// Author: Max
+//
+// This is an implementation of the MCS mutex described by Mellor-Crummey and Scott in their 1991 paper
+// “Algorithms for Scalable Synchronization on Shared-Memory Multiprocessors”
+//
+// This is a spinlock mutex, but unlike a simple spinlock in which all threads update and spin on
+// a single memory location, the MCS lock builds a linked-list of memory locations local to each core.
+//
+// Cores atomically append their local memory region to the global list using an unconditional
+// amoswap operation. They then spin on their local memories for a predecessor in the queue
+// to notify them that they now hold the lock.
+//
+// Once a core has completed its critical region, it checks for a successor and updates releases the lock to them.
+//
+// The advantages of this mutex over a simple spin lock on the manycore are two fold:
+//
+// (1) It greatly reduces the number of memory reqeusts on the network and it mitigates the extent to which
+// a single memory bank becomes a hot-spot. The number of requests issued to a memory bank containing the
+// lock object is linear with the number of times an acquire operation is execution.
+//
+// (2) The lock approximates a FIFO-ish structure, which improves fairness. A simple spinlock on the manycore
+// will favor threads topologically closer to the memory bank in which the lock resides and can lead to
+// starvation of the other cores.
+//
+// This lock is by no means perfect. For locks with low contention, a simple spinlock may result in better performance.
+
+        .text
+        .globl bsg_mcs_mutex_acquire
+bsg_mcs_mutex_acquire:
+        // a0 = mtx         : bsg_mcs_mutex_t*
+        // a1 = lcl         : bsg_mcs_mutex_node_t*
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*
+        sw      x0, 4(a1)                      // lcl->next = 0
+        sw      x0, 0(a1)                      // lcl->unlocked = 0
+        amoswap.w.aq t0, a2, 0(a0)             // predecessor = swap (&mtx, lcl_as_glbl)
+        beqz    t0, bsg_mcs_mutex_acquire_ret  // return if predecessor = 0
+        sw      a2, 4(t0)                      // predecessor->next = lcl_as_glbl
+        lr.w    t0, (a1)                       // unlocked = lcl->unlocked
+        bnez    t0, bsg_mcs_mutex_acquire_ret  // return if unlocked
+bsg_mcs_mutex_acquire_loop:
+        lr.w.aq t0, (a1)                       // unlocked = lcl->unlocked
+        beqz    t0, bsg_mcs_mutex_acquire_loop // while !unlocked
```

Comment:
// backwards predict not taken branch variant would help here

Reply:
// note that this is also polling, since it does not reset the reservation
// what is the expected wait time in this case?

Reply:
> note that this is also polling, since it does not reset the reservation
should it branch back to making the reservation, then?

> what is the expected wait time in this case?
the wait time can be arbitrarily long as it depends on both lock contention and the time required for the client's critical region. hopefully the client code does not require more than 100 cycles (possible if there's a cache miss) and contention is low (maybe no more than 2 cores waiting on the lock at once).

> is it possible that somebody would be writing to this register to do anything except unlockit?
memory location or register?
for the memory location: gosh I hope not. only two writes happen to this memory. the first is before a core tries to acquire the lock and it clears it. the second is when the lock owner is releasing the lock and unlocking its successor.
for the register: are we talking about a situation where we have context switched?